File: test-clients/csharp/SchemaTestRunner.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Weaviate.Client;

namespace WeaviateSchemaTest;

/// <summary>
/// Test runner for importing and exporting Weaviate schemas.
/// </summary>
public class SchemaTestRunner
{
    private readonly WeaviateClient _client;
    private readonly string _weaviateUrl;

    private SchemaTestRunner(WeaviateClient client, string weaviateUrl)
    {
        _client = client;
        _weaviateUrl = weaviateUrl;
    }

    /// <summary>
    /// Create a new SchemaTestRunner connected to local Weaviate.
    /// </summary>
    public static async Task<SchemaTestRunner> CreateAsync(string weaviateUrl = "localhost")
    {
        var client = await Connect.Local();
        return new SchemaTestRunner(client, weaviateUrl);
    }

    /// <summary>
    /// Load a schema from JSON file.
    /// </summary>
    public JObject LoadSchema(string schemaPath)
    {
        Console.WriteLine($"Loading schema from: {schemaPath}");
        var json = File.ReadAllText(schemaPath);
        return JObject.Parse(json);
    }

    /// <summary>
    /// Import a schema into Weaviate by creating a collection.
    /// </summary>
    public async Task<string> ImportSchema(JObject schema)
    {
        var collectionName = schema["name"]?.ToString();

        if (string.IsNullOrEmpty(collectionName))
        {
            throw new ArgumentException("Schema missing 'name' field");
        }

        Console.WriteLine($"Importing collection: {collectionName}");

        // Delete if exists
        try
        {
            var exists = await _client.Collections.Exists(collectionName);
            if (exists)
            {
                Console.WriteLine($"Deleting existing collection: {collectionName}");
                await _client.Collections.Delete(collectionName);
            }
        }
        catch
        {
            // Collection doesn't exist, which is fine
        }

        // Create the collection using JSON
        var jsonString = schema.ToString();
        await _client.Collections.CreateFromJson(jsonString);

        Console.WriteLine($"Successfully imported collection: {collectionName}");
        return collectionName;
    }

    /// <summary>
    /// Export a collection's schema configuration.
    /// </summary>
    public async Task<JObject> ExportSchema(string collectionName)
    {
        Console.WriteLine($"Exporting schema for collection: {collectionName}");

        var exportedConfig = await _client.Collections.Export(collectionName);

        if (exportedConfig == null)
        {
            throw new Exception($"Failed to export collection: {collectionName}");
        }

        // Convert CollectionConfigExport to JSON string then to JObject
        var jsonString = System.Text.Json.JsonSerializer.Serialize(exportedConfig);
        var exported = JObject.Parse(jsonString);

        Console.WriteLine($"Exported schema for collection: {collectionName}");
        return exported;
    }

    /// <summary>
    /// Save schema to JSON file.
    /// </summary>
    public void SaveSchema(JObject schema, string outputPath)
    {
        var directory = Path.GetDirectoryName(outputPath);
        if (!string.IsNullOrEmpty(directory))
        {
            Directory.CreateDirectory(directory);
        }

        File.WriteAllText(outputPath, schema.ToString(Formatting.Indented));

        Console.WriteLine($"Saved schema to: {outputPath}");
    }

    /// <summary>
    /// Delete a collection from Weaviate.
    /// </summary>
    public async Task Cleanup(string collectionName)
    {
        try
        {
            var exists = await _client.Collections.Exists(collectionName);
            if (exists)
            {
                await _client.Collections.Delete(collectionName);
                Console.WriteLine($"Cleaned up collection: {collectionName}");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Failed to cleanup collection: {ex.Message}");
        }
    }

    /// <summary>
    /// Run a complete import/export test.
    /// </summary>
    public async Task<TestResult> RunTest(string baselinePath, string outputDir, string schemaName)
    {
        var result = new TestResult(schemaName);

        try
        {
            // Load baseline
            var baseline = LoadSchema(baselinePath);

            // Import schema
            var collectionName = await ImportSchema(baseline);

            // Export schema
            var exported = await ExportSchema(collectionName);

            // Save exported schema
            var outputPath = Path.Combine(outputDir, "csharp", schemaName, "config.json");
            SaveSchema(exported, outputPath);

            result.Success = true;
            result.ExportedPath = outputPath;

            // Cleanup
            await Cleanup(collectionName);
        }
        catch (Exception ex)
        {
            result.Success = false;
            result.Error = ex.Message;
            Console.WriteLine($"Test failed for {schemaName}: {ex.Message}");
        }

        return result;
    }
}

/// <summary>
/// Result of a test run.
/// </summary>
public class TestResult
{
    public string SchemaName { get; }
    public bool Success { get; set; }
    public string? Error { get; set; }
    public string? ExportedPath { get; set; }

    public TestResult(string schemaName)
    {
        SchemaName = schemaName;
        Success = false;
    }
}

File: test-clients/java/src/main/java/io/weaviate/schema/test/SchemaTestRunner.java
package io.weaviate.schema.test;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.weaviate.client.Config;
import io.weaviate.client.WeaviateClient;
import io.weaviate.client.base.Result;
import io.weaviate.client.v1.schema.model.Property;
import io.weaviate.client.v1.schema.model.WeaviateClass;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Test runner for importing and exporting Weaviate schemas.
 */
public class SchemaTestRunner {
    private static final Logger logger = LoggerFactory.getLogger(SchemaTestRunner.class);

    private final WeaviateClient client;
    private final ObjectMapper objectMapper;
    private final String weaviateUrl;

    public SchemaTestRunner(String weaviateUrl) {
        this.weaviateUrl = weaviateUrl;
        Config config = new Config("http", weaviateUrl + ":8080");
        this.client = new WeaviateClient(config);
        this.objectMapper = new ObjectMapper();
    }

    /**
     * Load a schema from JSON file.
     */
    public JsonNode loadSchema(String schemaPath) throws IOException {
        logger.info("Loading schema from: {}", schemaPath);
        File file = new File(schemaPath);
        return objectMapper.readTree(file);
    }

    /**
     * Import a schema into Weaviate by creating a class.
     */
    public String importSchema(JsonNode schema) throws Exception {
        String className = schema.has("name") ? schema.get("name").asText() : null;

        if (className == null) {
            throw new IllegalArgumentException("Schema missing 'name' field");
        }

        logger.info("Importing class: {}", className);

        // Delete if exists
        Result<WeaviateClass> existsResult = client.schema().classGetter().withClassName(className).run();
        if (existsResult.getResult() != null) {
            logger.info("Deleting existing class: {}", className);
            client.schema().classDeleter().withClassName(className).run();
        }

        // Build WeaviateClass from JSON
        WeaviateClass weaviateClass = buildWeaviateClass(schema);

        // Create the class
        Result<Boolean> result = client.schema().classCreator().withClass(weaviateClass).run();

        if (result.hasErrors()) {
            throw new Exception("Failed to create class: " + result.getError());
        }

        logger.info("Successfully imported class: {}", className);
        return className;
    }

    /**
     * Build a WeaviateClass object from JSON schema.
     */
    private WeaviateClass buildWeaviateClass(JsonNode schema) {
        WeaviateClass.WeaviateClassBuilder builder = WeaviateClass.builder();

        // Set class name
        if (schema.has("name")) {
            builder.className(schema.get("name").asText());
        }

        // Set description
        if (schema.has("description")) {
            builder.description(schema.get("description").asText());
        }

        // Build properties
        if (schema.has("properties")) {
            List<Property> properties = new ArrayList<>();
            schema.get("properties").forEach(prop -> {
                Property.PropertyBuilder propBuilder = Property.builder()
                        .name(prop.get("name").asText());

                if (prop.has("dataType")) {
                    List<String> dataTypes = new ArrayList<>();
                    prop.get("dataType").forEach(dt -> dataTypes.add(dt.asText()));
                    propBuilder.dataType(dataTypes);
                }

                if (prop.has("description")) {
                    propBuilder.description(prop.get("description").asText());
                }

                properties.add(propBuilder.build());
            });
            builder.properties(properties);
        }

        // Set vectorizer
        if (schema.has("vectorizer")) {
            builder.vectorizer(schema.get("vectorizer").asText());
        } else {
            builder.vectorizer("none");
        }

        // Note: Java client has limited support for vectorConfig
        // This is a known limitation that the framework will expose

        return builder.build();
    }

    /**
     * Export a class's schema configuration.
     */
    public JsonNode exportSchema(String className) throws Exception {
        logger.info("Exporting schema for class: {}", className);

        Result<WeaviateClass> result = client.schema().classGetter().withClassName(className).run();

        if (result.hasErrors()) {
            throw new Exception("Failed to export schema: " + result.getError());
        }

        WeaviateClass weaviateClass = result.getResult();

        // Convert WeaviateClass to JsonNode
        String json = objectMapper.writeValueAsString(weaviateClass);
        JsonNode exported = objectMapper.readTree(json);

        logger.info("Exported schema for class: {}", className);
        return exported;
    }

    /**
     * Save schema to JSON file.
     */
    public void saveSchema(JsonNode schema, String outputPath) throws IOException {
        Path path = Paths.get(outputPath);
        Files.createDirectories(path.getParent());

        objectMapper.writerWithDefaultPrettyPrinter()
                .writeValue(path.toFile(), schema);

        logger.info("Saved schema to: {}", outputPath);
    }

    /**
     * Delete a class from Weaviate.
     */
    public void cleanup(String className) {
        try {
            Result<WeaviateClass> existsResult = client.schema().classGetter().withClassName(className).run();
            if (existsResult.getResult() != null) {
                client.schema().classDeleter().withClassName(className).run();
                logger.info("Cleaned up class: {}", className);
            }
        } catch (Exception e) {
            logger.warn("Failed to cleanup class: {}", e.getMessage());
        }
    }

    /**
     * Run a complete import/export test.
     */
    public TestResult runTest(String baselinePath, String outputDir, String schemaName) {
        TestResult result = new TestResult(schemaName);

        try {
            // Load baseline
            JsonNode baseline = loadSchema(baselinePath);

            // Import schema
            String className = importSchema(baseline);

            // Export schema
            JsonNode exported = exportSchema(className);

            // Save exported schema
            String outputPath = Paths.get(outputDir, "java", schemaName, "config.json").toString();
            saveSchema(exported, outputPath);

            result.setSuccess(true);
            result.setExportedPath(outputPath);

            // Cleanup
            cleanup(className);

        } catch (Exception e) {
            result.setSuccess(false);
            result.setError(e.getMessage());
            logger.error("Test failed for {}: {}", schemaName, e.getMessage(), e);
        }

        return result;
    }

    /**
     * Result of a test run.
     */
    public static class TestResult {
        private final String schemaName;
        private boolean success;
        private String error;
        private String exportedPath;

        public TestResult(String schemaName) {
            this.schemaName = schemaName;
            this.success = false;
        }

        public String getSchemaName() {
            return schemaName;
        }

        public boolean isSuccess() {
            return success;
        }

        public void setSuccess(boolean success) {
            this.success = success;
        }

        public String getError() {
            return error;
        }

        public void setError(String error) {
            this.error = error;
        }

        public String getExportedPath() {
            return exportedPath;
        }

        public void setExportedPath(String exportedPath) {
            this.exportedPath = exportedPath;
        }
    }
}
